In [63]:
# Importation des librairies

import numpy as np
import pandas as pd
import warnings

# Ignorer les warnings
warnings.filterwarnings('ignore')

# Fixer seed pour toujours avoir les mêmes valeurs aléatoires
np.random.seed(42)

features = ['ref_cli', 'prov', 'reg', 'com', 'fkt', 'date_dep_dmd', 'date_val_dmd',
            'date_int', 'tar', 'cat', 'pce']

cp_trn = np.arange(101, 120)
cp_tve = np.arange(501, 519)
cp_mjg = np.arange(401, 421)
cp_fnr = np.arange(301, 324)
cp_die = np.arange(202, 210)
cp_tul = np.arange(601, 622)

cp = dict()
cp['Antananarivo'] = cp_trn


# cp['Toamasina'] = cp_tve
# cp['Mahajanga'] = cp_mjg
# cp['Fianarantsoa'] = cp_fnr
# cp['Antsiranana'] = cp_die
# cp['Toliara'] = cp_tul


### Contrainte & conditions: 
    - Depuis 1980 à 2023
    - de janvier à décembre
    - Valeur aléatoire de 1 à 10 pour le nombre de demandes déposées pour chaque Fokontany
    - Saisonnalité 
    - Tendance progressive des données au fil du temps

In [64]:
# Importation de la liste des provinces, regions, districts, communes, fokontany de Madagascar
data_df = pd.read_csv('pro_reg_dist_com.csv', delimiter=';')
data_df = data_df.drop('N°', axis=1)

prod_conso_energy = pd.read_csv('scrap/prod_conso_energy.csv')
prod_conso_energy

Date  Production GWh  Consommation GWh  Consommation par habitant kWh
0   1980         417.000           388.000                            447
1   1981         420.000           391.000                            438
2   1982         431.000           401.000                            438
3   1983         452.000           420.000                            447
4   1984         476.000           443.000                            459
5   1985         483.000           449.000                            454
6   1986         482.000           448.000                            441
7   1987         478.000           445.000                            426
8   1988         490.000           456.000                            425
9   1989         541.000           503.000                            457
10  1990         574.000           534.000                            472
11  1991         568.000           528.000                            455
12  1992         580.000           539.000                            453
13  1993         581.000           540.000                            442
14  1994         584.000           543.000                            431
15  1995         609.000           566.000                            436
16  1996         664.000           618.000                            462
17  1997         695.000           646.000                            469
18  1998         750.000           698.000                            491
19  1999         777.000           723.000                            494
20  2000         780.000           747.000                            496
21  2001         886.000           848.000                            547
22  2002         866.000           827.000                            517
23  2003         990.000           945.000                            574
24  2004           1.100             1.054                            622
25  2005           1.111             1.064                            609
26  2006           1.134             1.089                            605
27  2007           1.184             1.131                            610
28  2008           1.232             1.182                            619
29  2009           1.235             1.173                            596
30  2010           1.329             1.260                            622
31  2011           1.348             1.271                            609
32  2012           1.701             1.617                            752
33  2013           1.856             1.762                            795
34  2014           1.970             1.873                            821
35  2015           2.052             1.949                            829
36  2016           2.024             1.916                            792
37  2017           2.211             2.101                            842
38  2018           2.271             2.166                            844
39  2019           2.020             1.889                            686
40  2020           1.976             1.862                            660
41  2021           2.093             1.956                            676

In [65]:
prod_conso_energy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 42 entries, 0 to 41
Data columns (total 4 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Date                           42 non-null     int64  
 1   Production GWh                 42 non-null     float64
 2   Consommation GWh               42 non-null     float64
 3   Consommation par habitant kWh  42 non-null     int64  
dtypes: float64(2), int64(2)
memory usage: 1.4 KB


In [66]:
# Ajout 2 nouvelles observation pour l'année 2022, 2023
line1 = pd.DataFrame([[2022, 215.0, 200.92, 694]], columns=prod_conso_energy.columns)
line2 = pd.DataFrame([[2023, 230.0, 218.55, 940]], columns=prod_conso_energy.columns)
prod_conso_energy = pd.concat([prod_conso_energy, line1])
prod_conso_energy = pd.concat([prod_conso_energy, line2])
prod_conso_energy = prod_conso_energy.reset_index(drop=True)

# Correction des valeurs de 'Consommation par habitant kWh'
prod_conso_energy['Consommation par habitant kWh'] = prod_conso_energy['Consommation par habitant kWh'].apply(lambda x: x/10)

# Correction des valeurs de 'Consommation GWh'
prod_conso_energy['Consommation GWh'] = prod_conso_energy['Consommation GWh'].apply(lambda x: x/1000 if x > 3 else x)

# Correction des valeurs de 'Production GWh'
prod_conso_energy['Production GWh'] = prod_conso_energy['Production GWh'].apply(lambda y: y/1000 if y > 3 else y)


prod_conso_energy

Date  Production GWh  Consommation GWh  Consommation par habitant kWh
0   1980           0.417           0.38800                           44.7
1   1981           0.420           0.39100                           43.8
2   1982           0.431           0.40100                           43.8
3   1983           0.452           0.42000                           44.7
4   1984           0.476           0.44300                           45.9
5   1985           0.483           0.44900                           45.4
6   1986           0.482           0.44800                           44.1
7   1987           0.478           0.44500                           42.6
8   1988           0.490           0.45600                           42.5
9   1989           0.541           0.50300                           45.7
10  1990           0.574           0.53400                           47.2
11  1991           0.568           0.52800                           45.5
12  1992           0.580           0.53900                           45.3
13  1993           0.581           0.54000                           44.2
14  1994           0.584           0.54300                           43.1
15  1995           0.609           0.56600                           43.6
16  1996           0.664           0.61800                           46.2
17  1997           0.695           0.64600                           46.9
18  1998           0.750           0.69800                           49.1
19  1999           0.777           0.72300                           49.4
20  2000           0.780           0.74700                           49.6
21  2001           0.886           0.84800                           54.7
22  2002           0.866           0.82700                           51.7
23  2003           0.990           0.94500                           57.4
24  2004           1.100           1.05400                           62.2
25  2005           1.111           1.06400                           60.9
26  2006           1.134           1.08900                           60.5
27  2007           1.184           1.13100                           61.0
28  2008           1.232           1.18200                           61.9
29  2009           1.235           1.17300                           59.6
30  2010           1.329           1.26000                           62.2
31  2011           1.348           1.27100                           60.9
32  2012           1.701           1.61700                           75.2
33  2013           1.856           1.76200                           79.5
34  2014           1.970           1.87300                           82.1
35  2015           2.052           1.94900                           82.9
36  2016           2.024           1.91600                           79.2
37  2017           2.211           2.10100                           84.2
38  2018           2.271           2.16600                           84.4
39  2019           2.020           1.88900                           68.6
40  2020           1.976           1.86200                           66.0
41  2021           2.093           1.95600                           67.6
42  2022           0.215           0.20092                           69.4
43  2023           0.230           0.21855                           94.0

In [67]:
# Nettoyage des données
df_copied = data_df.copy()

# Remplacement des valeurs '\\N' par null
df_copied.replace('\\N', np.nan, inplace=True)

# Remplacement des valeurs 'Unknown Name' par null
df_copied.replace('Unknown Name', np.nan, inplace=True)

# Suppression de toutes les lignes contenant les valeurs nulles
df_copied.dropna(inplace=True)

# Réinitialistion de l'index après suppression
fkt_df = df_copied.reset_index(drop=True)
fkt_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17121 entries, 0 to 17120
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   fokontany  17121 non-null  object
 1   Kaomina    17121 non-null  object
 2   Distrika   17121 non-null  object
 3   Region     17121 non-null  object
 4   Province   17121 non-null  object
dtypes: object(5)
memory usage: 668.9+ KB


D'après les informations affichées, il y a 17121 observations dans le data, lesquelles sont organisées en Provinces, Regions, Districts, Communes et Fokontany ou quartier.  
Mais ce qui nous interesse c'est de récupérer seulement la liste des districts de la région d'Analamanga de la province d'Antananarivo. L'idée c'est de travailler sur un échantillon un peu réduit pour éviter les calculs trop important.

In [68]:
# Filtre par Région d'Analamanga
analamanga_df = fkt_df[fkt_df['Region'] == 'Analamanga']
analamanga_df.reset_index(drop=True)

# Récupération de la liste de districts
districts = analamanga_df['Distrika'].unique()
districts

array(['Andramasina', 'Antananarivo-Avaradrano',
       'Antananarivo-Atsimondrano', 'Manjakandriana', 'Ambohidratrimo',
       'Anjozorobe', 'Ankazobe'], dtype=object)

In [69]:
# index_sep = prod_conso_energy[prod_conso_energy['Date'] == 2000].index[0]
# index_sep

# prod_conso_energy = prod_conso_energy[index_sep:]
# prod_conso_energy.reset_index(drop=True)
# prod_conso_energy.info()

In [70]:
# Création des données
years = range(1980, 2024)

data = {'district': [],
        'nb_dmd': [],
        'energy': [],
        'date_dep_dmd': []}

for year in years:
    # conso_mean_monthly = [prod_conso_energy['Consommation GWh'][prod_conso_energy.index == year]] / 12
    row = prod_conso_energy.loc[prod_conso_energy['Date'] == year]
    conso_mean_monthly = row.loc[row.index[0],'Consommation par habitant kWh'] / 12
    conso_mean_global = row.loc[row.index[0],'Consommation GWh'] / 12
    for month in range(1, 13):
        for district in districts:
            # Génération de la demande en électricité avec une saisonnalité et une tendance progressive
            demand = conso_mean_global + 10 * (year - 1980) + np.sin(month * np.pi / 6) * 20
            demand = int(demand)
            # Ajout d'une variation chaque année
            demand += np.random.randint(-20, 20)
            
            # Génération de la puissance énergétique avec une saisonnalité et une tendance progressive
            power = (conso_mean_monthly * demand) / 1_000_000
            
            data['district'].append(district)
            data['nb_dmd'].append(demand)
            data['energy'].append(power)
            data['date_dep_dmd'].append(pd.Timestamp(year, month, 1))

# Création du dataframe
df = pd.DataFrame(data)


In [75]:
# Modification du nom de la colonne
df.rename(columns={'date_dep_dmd':'date'}, inplace=True)
df[df['nb_dmd'] < 0]

district  nb_dmd    energy       date
3               Manjakandriana      -3 -0.000011 1980-01-01
30   Antananarivo-Atsimondrano      -8 -0.000030 1980-05-01
32              Ambohidratrimo      -4 -0.000015 1980-05-01
34                    Ankazobe      -2 -0.000007 1980-05-01
36     Antananarivo-Avaradrano      -3 -0.000011 1980-06-01
..                         ...     ...       ...        ...
239    Antananarivo-Avaradrano      -8 -0.000029 1982-11-01
307                   Ankazobe      -6 -0.000022 1983-08-01
311             Manjakandriana      -8 -0.000030 1983-09-01
317  Antananarivo-Atsimondrano      -7 -0.000026 1983-10-01
319             Ambohidratrimo      -7 -0.000026 1983-10-01

[80 rows x 4 columns]

**date_dep_dmd**: c'est la date de dépôt de la demande effectué par un potentiel client de la JIRAMA.

**district**: Nom des disctricts de la région d'Analamanga

**nb_dmd**: Nombre de demandes déposées

**energy**: Puissance minimal en energie électrique requise, en Mega Watt.

On sauvegarde les données aux format 'csv' pour une utilisation ultérieure

In [72]:
# Sauvegarde de la collection de données
df.to_csv('energy_antananarivo_collected.csv', index=False)